File: vortexdbg-android/src/main/java/com/vortexdbg/app/VortexClassLoader.java
package com.vortexdbg.app;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.List;

/**
 * Carregador das classes do APP rodando na JVM host (arquitetura A1).
 *
 * Recebe os artefatos `.class`/`.jar` extraídos do APK (via JEB ou dex2jar) e os
 * carrega num classloader dedicado — permitindo que o Vortex-DBG execute a lógica
 * Java real do app na JVM host (substituindo o fluxo LSPosed/Zygote num device).
 *
 * O parent padrão é o classloader do próprio Vortex, de modo que as classes do app
 * enxergam `java.*` (e, quando a camada de framework do WF4 estiver no classpath,
 * também `android.*`).
 */
public class VortexClassLoader extends URLClassLoader {

    public VortexClassLoader(ClassLoader parent, File... sources) {
        super(toUrls(sources), parent);
    }

    public VortexClassLoader(File... sources) {
        this(VortexClassLoader.class.getClassLoader(), sources);
    }

    private static URL[] toUrls(File[] sources) {
        List<URL> urls = new ArrayList<>(sources.length);
        for (File f : sources) {
            if (!f.exists()) {
                throw new IllegalArgumentException("source not found: " + f);
            }
            try {
                urls.add(f.toURI().toURL());
            } catch (MalformedURLException e) {
                throw new IllegalArgumentException(f.toString(), e);
            }
        }
        return urls.toArray(new URL[0]);
    }

    /** Carrega uma classe do app pelo nome binário (ex.: "org.cf.crypto.XORCrypt"). */
    public Class<?> loadApp(String binaryName) {
        try {
            return loadClass(binaryName);
        } catch (ClassNotFoundException e) {
            throw new IllegalStateException("app class not found: " + binaryName, e);
        }
    }
}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/app/VortexClassLoader.kt
package com.vortexdbg.app

import java.io.File
import java.net.MalformedURLException
import java.net.URL
import java.net.URLClassLoader

/**
 * Loader for the APP's classes running on the host JVM (A1 architecture).
 *
 * Takes the .class/.jar artifacts extracted from the APK (via JEB or dex2jar) and loads
 * them in a dedicated classloader — letting Vortex-DBG run the app's real Java logic on
 * the host JVM (replacing the LSPosed/Zygote flow on a device).
 *
 * The default parent is Vortex's own classloader, so app classes can see java.* (and,
 * once the WF4 framework layer is on the classpath, android.* too).
 */
open class VortexClassLoader(parent: ClassLoader?, vararg sources: File) :
    URLClassLoader(toUrls(sources), parent) {

    constructor(vararg sources: File) : this(VortexClassLoader::class.java.classLoader, *sources)

    /** Loads an app class by binary name (e.g. "org.cf.crypto.XORCrypt"). */
    open fun loadApp(binaryName: String): Class<*> {
        try {
            return loadClass(binaryName)
        } catch (e: ClassNotFoundException) {
            throw IllegalStateException("app class not found: $binaryName", e)
        }
    }

    companion object {
        private fun toUrls(sources: Array<out File>): Array<URL> {
            val urls = ArrayList<URL>(sources.size)
            for (f in sources) {
                if (!f.exists()) {
                    throw IllegalArgumentException("source not found: $f")
                }
                try {
                    urls.add(f.toURI().toURL())
                } catch (e: MalformedURLException) {
                    throw IllegalArgumentException(f.toString(), e)
                }
            }
            return urls.toTypedArray()
        }
    }
}
